### ISSUE TITLE
Stop using the term "official" about third-party drivers

### ISSUE THREAD

**Issue by mindplay-dk** [orientechnologies/orientdb#4354]
Your use of the term "official" to describe third-party open-source driver initiatives, is misleading, and doesn't serve you at all.

A competing product, [ArangoDB](https://www.arangodb.com/2015/06/performance-comparison-between-arangodb-mongodb-neo4j-and-orientdb/), just published a benchmark, in which OrientDB looks absolutely terrible, probably (at least in part) because they used the "official" JS driver, which still uses CSV-style serialization for transport.

When you label a driver "official", to most people, that's going to imply drivers that are supported, maintained, stable, performant, complete, up to date - and most people would expect they were developed or supervised by you, or your team.

If another competing company thinks they're using an "official" driver, you get results like this blog post.

I urge you to change the label "official" to something more accurate, like "approved" - or perhaps "certified", though that would imply to me that somebody from your team actually evaluated the drivers for quality and completeness.

To my knowledge, the only "official" drivers at the moment, that you support and develop, are the Java drivers, is that correct?

**Comment by mindplay-dk** [comment]
Incidentally, the only two "official" drivers on the drivers page *aren't* labeled "official".

**Comment by mindplay-dk** [comment]
Come to think of it, the very best term to use is probably simply "Recommended" - nothing more.

**Comment by smolinari** [comment]
Or just name them "3rd Party" or maybe "Community Sourced"? 

Scott

**Comment by sumpton** [comment]
Using oriento driver has been a very expensive and frustrating experience for us. We had expected the driver to be supported. The maintainer is looking for some significant cash to further develop it -- as I understand the situation.

Our lead developer could be more specific, but he has told me we can only use a small slice of OrientDB.

**Comment by sumpton** [comment]
We have been talking about this today and that the response appears to be to update the documentation instead of the Javascript driver. We have probably been very naive in believing that Javascript would become important and supported -- eventually. Our lead was suggesting that there isn't as much money for EE with Javascript and not important.

Oriento development has stalled. The maintainer has turned down casual contributors stating that the driver needs weeks of work and therefore a corporate sponsor. It doesn't look promising.

**Comment by smolinari** [comment]
Also, if a customer is buying into any support package, than usually support for "official" drivers would also be expected to be part of that support. If you aren't willing to do that, then the "official" title really should be dropped.

Scott

**Comment by mindplay-dk** [comment]
I was hoping to see driver support on other platforms getting ramped up, now that Luca Olivari is leading the charge - I think good driver support on all major fronts was a key factor to the success of MongoDB, and was hoping to see an actual official C or C++ driver, integrated with web platforms like PHP, Node and Ruby...

**Comment by lvca** [comment]
Hey guys, we are aware about the importance of drivers and we worked with each driver to unify efforts. We done and we're doing investments to hire people can support drivers internally or through partners. Unfortunately Oriento is an unfortunate case where the partnership didn't work. The good news is that in short time it will be supported directly by Orient Technologies.

**Comment by smolinari** [comment]
Is the PHP driver truly "official". This is the one I am very interested in actually and it says "official". :-)

Scott

**Comment by mindplay-dk** [comment]
> The good news is that in short time it will be supported directly by Orient Technologies.

That is very good news in deed! In other good news, [proper 64-bit integer support](https://wiki.php.net/rfc/size_t_and_int64_next) has finally arrived with the first [alpha release](http://php.net/archive/2015.php#id2015-01-11-6) of PHP7 yesterday.

Do you have a timeframe for the first official PHP7 driver beta?

Will this be a native PHP driver or an extension?

Will this driver be a custom API (like the current driver) or a PDO compatible driver?

Please keep me in mind, if you need beta testers - I have been waiting for this for years and I'd love to help out any way I can!

**Comment by sumpton** [comment]
Wonderful news regarding Node.

We started testing ArangoDB, reluctantly, because our lead likes and respects OrientDB very much, and thinks the fit is perfect for our project.

It is constructive to know you recognize the pain of communicating through Oriento, and care about JS.

**Comment by Ovidiu-S** [comment]
I can't speak for the Node community, but I for one am looking forward for an "official" JS driver to orientdb.

**Comment by smolinari** [comment]
I am with @mindplay-dk. I'd love to help out with any PHP driver from a testing perspective. 

If any new driver is designed for PHP(7), I would hope it is also a .so extension, like what MongoDB has and is also doing with their new driver. 

Scott

**Comment by mindplay-dk** [comment]
@sumpton we also have been evaluating ArangoDB, reluctantly, but turns out this database is "mostly in-memory", did you know that? It loads everything into RAM - it doesn't even materialize indexes on disk, only in-memory. So although it looks interesting, I can't see ArangoDB becoming a general-purpose database, e.g. no replacement for MySQL, which I could totally see OrientDB becoming. Once we have drivers, that is :-)

**Comment by phpnode** [comment]
@sumpton 

> The maintainer is looking for some significant cash to further develop it -- as I understand the situation.

Not the case at all, we just cannot justify actively developing a new major version of a driver for a database that none of our clients use any more.

> The maintainer has turned down casual contributors stating that the driver needs weeks of work and therefore a corporate sponsor. It doesn't look promising.

We've never turned down a reasonable contribution, there are 3 other independent contributors who have commit access to oriento. 

There is a lot of work already done on oriento v2, you can see it [here](https://github.com/codemix/oriento/pull/182), also there's a new query builder [here](https://github.com/codemix/oriento-query-builder). Anyone is welcome to pick up those projects and continue them, such is the nature of open source, but given that we have no incentive, it won't be us.

**Comment by smolinari** [comment]
Isn't this a chicken and egg problem too?

Scott

**Comment by CoDEmanX** [comment]
@mindplay-dk: ArangoDB loads and unloads the data as necessary to/from RAM (it uses memory-mapped files). Indexes are purely in-memory for highest-possible performance, but of course it would be great in future, if it could run in a "memory saving mode" as well to be suitable for commodity hardware. I would like to see indexes persisted on shutdown to avoid lengthy re-creation on server startup, although 2.6 will feature threaded index initialization and it shouldn't take long for moderate amounts of data.

**Comment by lvca** [comment]
Hi, did you see this news? http://orientdb.com/welcome-to-orientjs/

**Comment by sumpton** [comment]
We would have responded with, "fabulous" sooner, but we were doing the happy dance.

**Comment by lvca** [comment]
:-)

**Comment by mindplay-dk** [comment]
Great news for Node users! I'm stuck with PHP. I remain hopeful :-)

**Comment by smolinari** [comment]
Why stuck with PHP? The PHP driver is also officially supported. 

Scott

**Comment by mindplay-dk** [comment]
> The PHP driver is also officially supported

see above ;-)

**Comment by lvca** [comment]
The situation with PHP driver is different. Even if it hasn't been adopted internally the committers are very available to fix things and so far we didn't receive complaints about it. @mindplay-dk What's your problem with PHP driver?

**Comment by a-unite** [comment]
People, hi!
I'm the one who doesn't care about labels, but about reality. Oriento was one of the main reasons, why we started with OrientDB. It was and still is the only _really_ usable driver for non Java languages here.

Moreover, Charles Pick is the man, who (without any special intention from his side) convinced me to pick up ODB.
When you are working with open-source, just rising, not very popular, but promising project - you have to agree that some inconsistency and problems will happen while in active development stage.
But what is more important - you choose community around the project, not only product (growth is what makes start-up valuable), after all.

And @phpnode  is the one of the best common sense providers here (I mean not only for Oriento, but for OrientDB too).
Instead of blaming him for his inability to spend few weeks/months for free, and talk about naming, I hope Orient Technologies (with our help, hopefully) could find the way to involve him on even deeper partnership.

**Comment by lvca** [comment]
@a-unite Actually Charles's company was an official partner... Nobody blames him to have no time to work on the project, most of the contributor of OrientDB works in their spare time and no one can say to be pushed/blamed on anything. The problem here was completely different. Unfortunately very very very bad things happened. However, I don't think that sharing what's happened on an issue on GitHub is a good idea. OrientDB is an open project, made of a great engine and other pieces that are around it, like drivers, plugins, etc. We work so hard to make OrientDB the best DBMS ever. We work with passion on it and sometimes we found amazing people that contribute to it. But I know everybody can't be happy. Somebody can have different ideas, somebody can act like a partner, then change his mind, deciding to go with other NoSQL vendors, talk bad about your product publicly and even be caught suggesting to OrientDB clients to use other products... 

We really appreciate the big work by all the contributors behind Oriento and we hope they will continue on working on OrientJS with the same passion.

I'm going to close this issue, I think we already talked too much about this ;-)

**Comment by dmarcelino** [comment]
I actually agree with @a-unite. Oriento is a great driver and in great measure this is due to @phpnode's tremendous work. It was because of Oriento that I've started using OrientDB and it is the driver that made [sails-orientdb](https://github.com/appscot/sails-orientdb) possible.

But Oriento is more than @phpnode. For instance it currently has 5 owners and 3 collaborators, forking it under a different name and npm package is effectively disowning this people. Forking a project is the last option and it actually goes against the current trend in big FOSS projects: io.js is merging back with node.js and lodash will be merged back to underscore.

I believe any of the following options would be better:

1. Issue PRs against Oriento (reasonable PRs have never been rejected);
2. Ask write access to Oriento repo;
3. Ask for Oriento project to be transferred over to Orient technologies.

Communities are important and any of these options would preserve the current existing community around Oriento, instead of fragmenting it.

Disclaimer: I'm one of the Oriento collaborators

**Comment by lvca** [comment]
@dmarcelino We kindly asked the transfer of "oriento" project under Orient Technologies umbrella (like we did with other drivers), but @phpnode asked for money... Is this the FOSS spirit you mentioned earlier? This is the main reason why we forked and changed name, to be able to move the driver forward with no treats.

**Comment by phpnode** [comment]
@lvca 

>  caught suggesting to OrientDB clients to use other products...

"caught" implies that I was doing something wrong which is absolutely not the case. What specifically happened was that someone said "I need graph database that supports transactions" and I suggested an alternative product because at that time OrientDB had no transaction support. You might not like that suggestion but that does not make it wrong. 

You're upset that I chose to speak honestly about the shortcomings of your product rather than sticking to the "orientdb is awesome" script. You would much rather silence dissent than fix the problems.

> We kindly asked the transfer of "oriento" project under Orient Technologies umbrella 

You didn't, you asked for the npm package, not the github projects at all. Since oriento was still being developed until literally [today](https://github.com/codemix/oriento/commits/master) (no, those commits aren't in orientjs) it made no sense for me to transfer the project to people who are not substantially involved in the project development. I did however assure you that any contributions you made would be reviewed and merged quickly, there was no blocker to ongoing collaboration from the oriento side. 


>  @phpnode asked for money... Is this the FOSS spirit you mentioned earlier? 

Careful now, firstly I didn't ask you for money, I said "make me an offer" which is quite different, but more importantly you seem to be conflating FOSS with "give me your work for free". There's a big difference here. I do not *owe* you the oriento project, it's something that me and the other collaborators put a lot of effort into, charged you nothing for and your company benefited significantly from. As open source contributors, we rarely benefit financially from our open source work directly, the benefit comes from more indirect links - publishing our work gives us exposure that helps us find new clients, attract developer talent, find more interesting work etc. If you want to claim that project for yourself, then you remove those benefits from the original author, therefore it is reasonable for them to expect something in exchange.

The fact that you *expect* me to hand over the project to you *in exchange for nothing*, while you write derogatory blog posts and call me out by name in public, whilst slapping a new name on my work and calling it your own, really shows your true nature.

**Comment by seeden** [comment]
At first I did not want to write anything but it looks like a small war. I understand both of you @lvca and @phpnode, you are upset a little bit. I really understand you because I know what happend. But please take care of us (developers). We need to have a good javascript driver. Ideal solution is to have only one otherwise it will split developers and pull requests too.  @lvca I know it looks very simple to fork this project. But are you sure that you are able to write a really good code in javascript? I saw the improvement of @phpnode in last half of the year. He is familiar with the latest ES6, babel and webpack. I saw that he understands what is the main problem of the combination ES6 with babel and he knows how to improve the performance as well. Please take breath, forget you are upset and try to find solution which will be good for community of the javascript developers.

**Comment by lvca** [comment]
These are the facts:
- On this topic you said "Anyone is welcome to pick up those projects and continue them, such is the nature of open source, but given that we have no incentive, it won't be us." (https://github.com/orientechnologies/orientdb/issues/4354#issuecomment-111859860)
- The main author and owner of oriento talk bad about OrientDB on many public places
- You claim he has no incentive to work on the driver anymore (I'm not interested about your reasons) 
- Last release you published on NPM is the 1.1.3 of about *4 months ago*, even if the project received a lot of contributions in the last 4 months (https://github.com/codemix/oriento/commits/master). Some contribution was made by OrientDB Team like the Live Query
- When we asked the ownership of the project (oriento, not only npm) you rejected it asking for "make an offer"

That said we get the decision to fork and rename the driver to avoid depending by you for all the reasons above. However we kept the history, so all the contributions have its authors. As you can see there is nothing about "claim that project for yourself, then you remove those benefits from the original author, therefore it is reasonable for them to expect something in exchange". All the contributors can continue on this driver like before, but in a project not lead by somebody that publicly turned against OrientDB project.

**Comment by phpnode** [comment]
https://github.com/orientechnologies/orientjs/commit/f1c08128c8569b82fd8a6729ba1bea09ec793a15#diff-b9cfc7f2cdf78a7f4b91a753d10865a2L16

**Comment by lvca** [comment]
This is the history of Oriento: https://github.com/codemix/oriento#history:
```
In 2012, Gabriel Petrovay created the original node-orientdb library, with a straightforward callback based API.

In early 2014, Giraldo Rosales made a whole host of improvements, including support for orientdb 1.7 and switched to a promise based API.

Later in 2014, codemix refactored the library to make it easier to extend and maintain, and introduced an API similar to nano. The result is so different from the original codebase that it warranted its own name and npm package. This also gave us the opportunity to switch to semantic versioning.
```

So why did you put your name as author if the original one was Gabriel Petrovay and then Giraldo Rosales? As you can see we did exactly what you did in the history of OrientJS:
https://github.com/orientechnologies/orientjs#history

**Comment by lvca** [comment]
https://github.com/codemix/oriento/blob/master/package.json#L17

**Comment by phpnode** [comment]
@lvca because not a single line of the code remains from those original authors. It's a from scratch rewrite that was done piece by piece, so it preserves the history. This is a trivial point however, but let's be clear, Enrico hasn't even written *one* line.

**Comment by lvca** [comment]
This is absolutely false. It's all public, so take a look at:
- oriento: https://github.com/codemix/oriento/blob/master/lib/utils.js#L80
- and node-orientdb: https://github.com/nitrog7/node-orientdb/blob/master/lib/utils.js#L69

Looks like identical. So your statement is false again "not a single line of the code remains from those original authors. It's a from scratch rewrite that was done piece by piece, so it preserves the history."

**Comment by phpnode** [comment]
@lvca I wrote that line, look at the history.

**Comment by seeden** [comment]
Is any chance to have only one good driver?

**Comment by StarpTech** [comment]
@lvca this is not the right way. Your way is a kind of opensource but it signalized the wrong way to work with a community. Where does the hate come from? The prefer way is the community way so create a PR or issue in oriento if you want to change something like contributors of orientdb.

**Comment by lvca** [comment]
@StarpTech There is no hate here. Users asked for better support for the Node.js driver, but how can we do that if we can't even release a new version on NPM?

**Comment by lvca** [comment]
About forking: http://www.techrepublic.com/article/why-you-should-fork-your-next-open-source-project/

**Comment by phpnode** [comment]
>  how can we do that if we can't even release a new version on NPM?

You could send a polite email saying "please can you publish a new version of oriento on npm". Or open an issue like anyone else would. Since you have had no involvement in oriento development you don't have the insight to even judge whether something is a breaking change or just a minor version bump etc, e.g. you released orientjs at 2.0.0 even though it wasn't a major bump, it's more appropriate for someone like @dmarcelino to have access.

**Comment by phpnode** [comment]
> somebody that publicly turned against OrientDB

What I find really interesting here is that you use this as a kind of insult with which to discredit me. You're not interested in acknowledging *why* someone who was so enthusiastic about your product and has spent so much time with it has grown so dissatisfied. Of course you'd rather I shut up, but Open Source isn't just about code, it's about open discussions. If OrientDB is an open project that welcomes contributions then it should welcome them even when they come in the form of negative feedback. If your first instinct is to attack the person making the claims, isn't that evidence of a deeper problem?

**Comment by smolinari** [comment]
I am just a bystander to the Oriento problem and from my probably not totally informed point of view, it seems there is an incorrect leveraging of the past and future development of Oriento for personal or commercial gains (whatever "make me an offer" or "given that we have no incentive, it won't be us" means). That is not what OSS is all about. As the lead of the software, either you actively support and contribute to it and can, or you don't. You most definitely should not hold the development up for ransom, which seems to be the issue here, both for the user wanting it to be further developed and for OrientDB, who wants to take over the lead to further develop it. 

I would like to suggest you change your perspective on what it means and how to earn from your hard work. No doubt, you should. It just seems, you are just going about it the wrong way. 

Scott

**Comment by dmarcelino** [comment]
Guys, **community** is about **collaboration** and collaboration does not necessarily means **development**. @phpnode has stated that he has no incentive to continue developing Oriento but he has up until now actively responded to **issues** and **merged PRs**. So yes, @phpnode is still actively collaborating and maintaining Oriento, contributing with his expertise in the process. This takes time and as we all know time is precious. Not only that, there are more people with write access to approve PRs in Oriento which proves development is not held to ransom by a single person.

@lvca, as the CEO of Orient I'm sure you want the best for OrientDB and its community, in this case the OrientDB Node.js community. This community revolves around Oriento and it's happy with how it has worked and it will be beneficial to keep that as it is. I believe it's in all's best interest to keep using Oriento as the official driver and for Orient Technologies to contribute to it. I understand you wanted more control but the truth is @phpnode has assured he will keep reviewing and merging PRs and the recent example of livequeries (codemix/oriento#306) proves that.

In other words there is a path open for Orient Technologies to contribute to Oriento and keep the community intact. Can we please try that before resorting to extreme measures as forking? If this collaboration does not work because of @phpnode then you'll have the community backing to fork or take over Oriento.

**Comment by phpnode** [comment]
@smolinari 

> probably not totally informed 

You're right, you're uninformed. No one is holding the project to ransom, oriento is still being developed, still accepting contributions, and has active maintainers. if you want to contribute to it, go ahead. If the goal is to deliver a quality driver that is well supported there's a path towards that which does not involve running slipshod all over the community and shitting on people's work.

**Comment by smolinari** [comment]
I hope you all can come to some amicable solution. 

Scott

**Comment by phpnode** [comment]
me too, believe it or not *I want to be able to use OrientDB*

**Comment by StarpTech** [comment]
@dmarcelino absolutely. First try and when it won't work a fork is the choice.

**Comment by lucaolivari** [comment]
I'm the President of Orient Technologies and have been part of the private discussion with @phpnode and on taking the decision regarding supporting the node.js driver directly. 

The Oriento driver came to life succeeding two previous versions (see [history](https://github.com/orientechnologies/orientjs#history)). We've been informed of the decision of the actual maintainer to no longer work on the next version of Oriento and that was publicly shared too. Everyone is free to decide wether they want to contribute or not and we respect their choice, but we felt the need to continue development and support.

We're still offering the driver under the previous license and welcome the previous committers and new ones to contribute. @phpnode and all the others still maintain the privilege of committers and we're looking forward to their contribution which will be a proof of their willingness to support the community further! On our end, we've devoted resources to actively develop the new official driver and encourage people to use it and the community to contribute. 

Let's continue to build on the excellent work that's been done by @phpnode lately and Giraldo and Gabriel before. 31 other committers also deserve a lot of credit. Thank you!

**Comment by phpnode** [comment]
@lucaolivari this is exactly the kind of shit-eating corporate doublespeak that makes people distrust you. I told you (and repeatedly clarified) that I would not be continuing development on the next *major* version of oriento (a full rewrite), but that the existing version would continue to be maintained by myself and the other contributors, and that we were happy to accept your contributions. 

It's a testament to your dishonesty that you're trying to spin this fork as anything other than the hostile, anti-community, blame-shirking behaviour it is.

**Comment by lucaolivari** [comment]
@phpnode this is not really productive or professional and just confirms what I said above. People need to rely on a driver that's supported now and for the long term and that's why we will continue supporting OrientJS going forward. Your contributions are welcome.

**Comment by phpnode** [comment]
Only in your world is the word *professional*  a synonym for *sycophant* or *weasel*.

As a professional, my clients come to me because I provide *honesty*. Sometimes part of being honest is saying things that other people don't like. This is the price of integrity.

**Comment by dmarcelino** [comment]
@lucaolivari, your comments makes me think that you either not read or completely disregarded my comment on https://github.com/orientechnologies/orientdb/issues/4354#issuecomment-113085248 and other comments from other people.

I find your decision and @lvca's very disappointing. Not only you are making an unilateral decision regarding the future of the Node.js driver you are also turning your backs to the very community that has supported and nurtured the Node.js driver in the last years. OrientJS may be Open Source but it's born of a *hostile takeover* and is not the *FOSS way* of doing things. Requiring full control of a project in order to contribute and support it is not how open communities work and thrive. This sort of behaviour discourages contributions and alienates people.

I'm genuinely sad with your decisions regarding this matter as it shows you are not willing to compromise with the community for the greater good.

**Comment by phpnode** [comment]
I've just added @dmarcelino as a collaborator for oriento on npm - he's a neutral third party that knows the project well. He has the same level of access to oriento that I do. He can publish new releases and has the ability to appoint new collaborators after they've proved themselves capable. If there was ever any blocker to ongoing oriento development, there isn't now.

**Comment by dmarcelino** [comment]
Oriento v1.2.0 [released](https://www.npmjs.com/package/oriento).

@lvca and @lucaolivari, I understand that you guys don't get along with @phpnode, I'm not asking for you to change your opinion about him, but like I said earlier Oriento is more than @phpnode and he clearly has signalled that he is willing share control over it. He won't and can't hold anybody to ransom so will you reconsider your decision?

I would like to keep this a productive and constructive discussion so we can bridge whatever differences that may subside.

**Comment by sumpton** [comment]
What does it matter if OrientDB fork the project and assume the lead role? Couldn't all the contributors shift as well?

Of course I would like Node to be a first class driver for OrientDB.

It appears the driver likely needs refactoring and it would be brilliant if the incredibly responsive core team were also thinking Node when they enhance the database that we and others have bet our projects on.

I don't understand any reluctance to shift the project the OrientDB Gods. I can think of lots of reasons why this should be better for our community.

**Comment by phpnode** [comment]
@sumpton sorry but you're drinking their koolaid, these people are manipulative liars - here's another example where @lvca impersonated a (famous) person to shill on behalf of orientdb - https://github.com/weinberger/nosql-tests/pull/1#issuecomment-113224861

This is a *database* vendor. If you cannot trust the honesty of their actions, what makes you think you can trust their technical claims?
